platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,203 @@
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Read-write interface for accessing rosters states.
+ */
+public class WritableRosterStore implements RosterStateModifier {
+
+    private final WritableStates writableStates;
+    private final RosterStateAccessor rosterStateAccessor;
+
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * The maximum number of active rosters to keep in the roster state.
+     */
+    static final int MAXIMUM_ROSTER_HISTORY_SIZE = 2;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        final Bytes incomingCandidateRosterHash =
+                RosterUtils.hash(candidateRoster).getBytes();
+
+        // update the roster state/map
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Bytes previousCandidateRosterHash = previousRosterState.candidateRosterHash();
+        final Builder newRosterState =
+                previousRosterState.copyBuilder().candidateRosterHash(incomingCandidateRosterHash);
+        removeRoster(previousCandidateRosterHash);
+        storeRoster(candidateRoster, incomingCandidateRosterHash, newRosterState);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
```

Comment:
Same here, the `rosterStateAccessor` is not aware of any changes until committed.

Reply:
https://github.com/hashgraph/hedera-services/pull/15442#discussion_r1805516449

Reply:
This has been resolved.